File: src/local_rag/parsers/epub.py
"""EPUB text extraction using zipfile + BeautifulSoup."""

import logging
import xml.etree.ElementTree as ET
import zipfile
from pathlib import Path, PurePosixPath

from bs4 import BeautifulSoup

logger = logging.getLogger(__name__)


def parse_epub(path: Path) -> list[tuple[int, str]]:
    """Extract text from an EPUB file, chapter by chapter.

    Opens the EPUB as a ZIP archive, reads the OPF manifest to find the
    spine order, then extracts text from each XHTML content document.

    Args:
        path: Path to the EPUB file.

    Returns:
        List of (chapter_number, text) tuples (1-based),
        same shape as parse_pdf() for consistency.
    """
    if not path.exists():
        logger.error("EPUB file not found: %s", path)
        return []

    try:
        return _extract_chapters(path)
    except zipfile.BadZipFile:
        logger.error("EPUB is not a valid ZIP archive: %s", path)
        return []
    except Exception as e:
        logger.error("Failed to parse EPUB %s: %s", path, e)
        return []


def _extract_chapters(path: Path) -> list[tuple[int, str]]:
    """Extract chapter text from an EPUB file."""
    with zipfile.ZipFile(path, "r") as zf:
        # Step 1: Find the OPF file via META-INF/container.xml
        opf_path = _find_opf_path(zf)
        if not opf_path:
            logger.warning("Could not find OPF file in EPUB: %s", path)
            return []

        # Step 2: Parse OPF to get spine item hrefs
        opf_dir = str(PurePosixPath(opf_path).parent)
        spine_hrefs = _parse_opf_spine(zf, opf_path)
        if not spine_hrefs:
            logger.warning("No spine items found in EPUB: %s", path)
            return []

        # Step 3: Extract text from each spine item
        chapters: list[tuple[int, str]] = []
        for chapter_num, href in enumerate(spine_hrefs, start=1):
            # Resolve href relative to OPF directory
            if opf_dir and opf_dir != ".":
                full_path = f"{opf_dir}/{href}"
            else:
                full_path = href

            try:
                content = zf.read(full_path).decode("utf-8", errors="replace")
            except KeyError:
                logger.debug("Spine item not found in archive: %s", full_path)
                continue

            text = _html_to_text(content)
            if text.strip():
                chapters.append((chapter_num, text.strip()))

        if not chapters:
            logger.warning("No extractable text found in EPUB %s (may be DRM-protected)", path)

        return chapters


def _find_opf_path(zf: zipfile.ZipFile) -> str | None:
    """Find the OPF file path from META-INF/container.xml."""
    try:
        container_xml = zf.read("META-INF/container.xml").decode("utf-8")
    except KeyError:
        return None

    root = ET.fromstring(container_xml)
    # Handle namespace
    ns = {"container": "urn:oasis:names:tc:opendocument:xmlns:container"}
    rootfile = root.find(".//container:rootfile", ns)
    if rootfile is not None:
        return rootfile.get("full-path")

    # Fallback: try without namespace
    for elem in root.iter():
        if elem.tag.endswith("rootfile"):
            return elem.get("full-path")

    return None


def _parse_opf_spine(zf: zipfile.ZipFile, opf_path: str) -> list[str]:
    """Parse the OPF file and return ordered list of content document hrefs from the spine."""
    try:
        opf_content = zf.read(opf_path).decode("utf-8")
    except KeyError:
        return []

    root = ET.fromstring(opf_content)

    # Detect OPF namespace
    ns_opf = ""
    if root.tag.startswith("{"):
        ns_opf = root.tag.split("}")[0] + "}"

    # Build manifest: id -> href
    manifest: dict[str, str] = {}
    for item in root.iter(f"{ns_opf}item"):
        item_id = item.get("id", "")
        href = item.get("href", "")
        media_type = item.get("media-type", "")
        if item_id and href:
            manifest[item_id] = (href, media_type)

    # Read spine order
    spine_hrefs: list[str] = []
    for itemref in root.iter(f"{ns_opf}itemref"):
        idref = itemref.get("idref", "")
        if idref in manifest:
            href, media_type = manifest[idref]
            # Only include XHTML/HTML content documents
            if "html" in media_type or "xml" in media_type:
                spine_hrefs.append(href)

    return spine_hrefs


def _html_to_text(html_content: str) -> str:
    """Convert HTML/XHTML content to plain text."""
    soup = BeautifulSoup(html_content, "html.parser")
    text = soup.get_text(separator="\n")
    # Clean up excessive blank lines
    lines = [line.strip() for line in text.splitlines()]
    return "\n".join(line for line in lines if line)

File: local-rag-go/internal/parser/epub.go
package parser

import (
	"archive/zip"
	"encoding/xml"
	"fmt"
	"log/slog"
	"path"
	"strings"
)

// ChapterText represents a single chapter of extracted EPUB text.
type ChapterText struct {
	ChapterNumber int // 1-based
	Text          string
}

// ParseEPUB extracts text from an EPUB file, chapter by chapter.
func ParseEPUB(filePath string) []ChapterText {
	reader, err := zip.OpenReader(filePath)
	if err != nil {
		slog.Error("failed to open EPUB", "path", filePath, "err", err)
		return nil
	}
	defer reader.Close()

	opfPath := findOPFPath(reader)
	if opfPath == "" {
		slog.Warn("could not find OPF file in EPUB", "path", filePath)
		return nil
	}

	opfDir := path.Dir(opfPath)
	spineHrefs := parseOPFSpine(reader, opfPath)
	if len(spineHrefs) == 0 {
		slog.Warn("no spine items found in EPUB", "path", filePath)
		return nil
	}

	var chapters []ChapterText
	for chapterNum, href := range spineHrefs {
		fullPath := href
		if opfDir != "." && opfDir != "" {
			fullPath = opfDir + "/" + href
		}

		content, err := readZipFile(reader, fullPath)
		if err != nil {
			slog.Debug("spine item not found in archive", "item", fullPath)
			continue
		}

		text := HTMLToText(content)
		if strings.TrimSpace(text) != "" {
			chapters = append(chapters, ChapterText{
				ChapterNumber: chapterNum + 1,
				Text:          strings.TrimSpace(text),
			})
		}
	}

	if len(chapters) == 0 {
		slog.Warn("no extractable text found in EPUB", "path", filePath)
	}

	return chapters
}

// container.xml structures
type containerXML struct {
	Rootfiles struct {
		Rootfile []struct {
			FullPath string `xml:"full-path,attr"`
		} `xml:"rootfile"`
	} `xml:"rootfiles"`
}

func findOPFPath(reader *zip.ReadCloser) string {
	content, err := readZipFile(reader, "META-INF/container.xml")
	if err != nil {
		return ""
	}

	var container containerXML
	if err := xml.Unmarshal([]byte(content), &container); err != nil {
		slog.Warn("failed to parse container.xml", "err", err)
		return ""
	}

	if len(container.Rootfiles.Rootfile) > 0 {
		return container.Rootfiles.Rootfile[0].FullPath
	}
	return ""
}

// OPF structures
type opfPackage struct {
	Manifest struct {
		Items []struct {
			ID        string `xml:"id,attr"`
			Href      string `xml:"href,attr"`
			MediaType string `xml:"media-type,attr"`
		} `xml:"item"`
	} `xml:"manifest"`
	Spine struct {
		Itemrefs []struct {
			IDRef string `xml:"idref,attr"`
		} `xml:"itemref"`
	} `xml:"spine"`
}

func parseOPFSpine(reader *zip.ReadCloser, opfPath string) []string {
	content, err := readZipFile(reader, opfPath)
	if err != nil {
		return nil
	}

	var pkg opfPackage
	if err := xml.Unmarshal([]byte(content), &pkg); err != nil {
		slog.Warn("failed to parse OPF", "err", err)
		return nil
	}

	// Build manifest: id -> (href, media-type)
	type manifestItem struct {
		href      string
		mediaType string
	}
	manifest := make(map[string]manifestItem)
	for _, item := range pkg.Manifest.Items {
		if item.ID != "" && item.Href != "" {
			manifest[item.ID] = manifestItem{href: item.Href, mediaType: item.MediaType}
		}
	}

	var spineHrefs []string
	for _, itemref := range pkg.Spine.Itemrefs {
		item, ok := manifest[itemref.IDRef]
		if !ok {
			continue
		}
		if strings.Contains(item.mediaType, "html") || strings.Contains(item.mediaType, "xml") {
			spineHrefs = append(spineHrefs, item.href)
		}
	}

	return spineHrefs
}

func readZipFile(reader *zip.ReadCloser, name string) (string, error) {
	for _, f := range reader.File {
		if f.Name == name {
			rc, err := f.Open()
			if err != nil {
				return "", err
			}
			defer rc.Close()
			var sb strings.Builder
			buf := make([]byte, 4096)
			for {
				n, err := rc.Read(buf)
				if n > 0 {
					sb.Write(buf[:n])
				}
				if err != nil {
					break
				}
			}
			return sb.String(), nil
		}
	}
	return "", fmt.Errorf("file not found in archive: %s", name)
}
